internal/controller/pkg/revision/runtime_function.go

```diff
@@ -115,16 +115,37 @@ func (h *FunctionHooks) Post(ctx context.Context, pkg runtime.Object, pr v1.Pack
 		return nil
 	}
 
-	sa := build.ServiceAccount()
-	if err := h.client.Apply(ctx, sa); err != nil {
-		return errors.Wrap(err, errApplyFunctionSA)
+	saName := build.ReferencedServiceAccountName()
+	// We only create the service account if it's not referenced in the
+	// DeploymentRuntimeConfig. This is to support the case where the
+	// ServiceAccount is managed externally.
+	if saName == "" {
+		sa := build.ServiceAccount()
+		saName = sa.Name
+		if err := h.client.Apply(ctx, sa); err != nil {
+			return errors.Wrap(err, errApplyFunctionSA)
+		}
 	}
 
-	d := build.Deployment(sa.Name, functionDeploymentOverrides(functionMeta, pr)...)
+	d := build.Deployment(saName, functionDeploymentOverrides(functionMeta, pr)...)
 	if err := h.client.Apply(ctx, d); err != nil {
 		return errors.Wrap(err, errApplyFunctionDeployment)
 	}
 
+	// Set the service account reference on the function revision. This is
+	// used by the RBAC manager to bind the generated ClusterRole to this
+	// service account.
+	// Persist it only if it's not set or if it has changed.
+	if saRef := pr.GetServiceAccountRef(); saRef == nil || (saRef.Name != saName || saRef.Namespace != d.Namespace) {
+		pr.SetServiceAccountRef(&v1.ServiceAccountReference{
+			Name:      saName,
+			Namespace: d.Namespace,
+		})
+		if err := h.client.Apply(ctx, pr); err != nil {
+			return errors.Wrap(err, "cannot apply function revision with service account reference")
+		}
+	}
```

Comment:
I find it a little surprising/unintuitive that this ref is set by the post hook as part of processing a runtimeconfig. I generally think of information as flowing from revision spec -> generated manifests. I suppose this manifests -> revision spec flow does match how we set TLS certificate names there today.

Reply:
Agreed. I am not happy with this either, but this seems to be better than the other alternative. 

When I try to move this to the same place where we set TLS certificate names, which is a package reconciler,
- I needed to duplicate the places where we're reading the DeploymentRuntimeConfig.
- I needed to make the package manager aware of the deployment namespace (to set SA namespace), which is only a runtime concept and currently only used by RuntimeHooks.
- I still needed to set `revision.Spec.ServiceAccountRef` here if no existing one is configured and we're building/creating a new one. 

I think the main difference here is, that we're not only using this field to transfer information only from "package -> revision" but also "revision -> rbac manager". So, when I try to move things up for the first part, I still need to persist in the service account (when we're creating it) for the second part. So, I would prefer having it in one place rather than duplicating logic code in two places.

Reply:
Another idea: does the provider deployment's `spec.template.spec.serviceAccountName` always point to the SA that we should bind roles to?

If so, could the RBAC manager find the SA it wants to operate on by listing deployments owned by the provider, and  granting access to the `serviceAccountName` they reference? That would let us avoid having to add this new field to the revision type.

Reply:
Thanks, I like this as this enables fixing the issue with less diff and checking the service account from the Deployment makes  much sense (e.g. I would do that if I need to figure out the SA being used for a provider revision).

Updated accordingly.